<b>Koushal Yeginwar
<li>Lab- Boston Assignment
<li>netid- kyegi2

In [1]:
from sklearn import datasets
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
import math

In [2]:
bean = datasets.load_boston()

In [3]:
print bean

{'data': array([[  6.32000000e-03,   1.80000000e+01,   2.31000000e+00, ...,
          1.53000000e+01,   3.96900000e+02,   4.98000000e+00],
       [  2.73100000e-02,   0.00000000e+00,   7.07000000e+00, ...,
          1.78000000e+01,   3.96900000e+02,   9.14000000e+00],
       [  2.72900000e-02,   0.00000000e+00,   7.07000000e+00, ...,
          1.78000000e+01,   3.92830000e+02,   4.03000000e+00],
       ..., 
       [  6.07600000e-02,   0.00000000e+00,   1.19300000e+01, ...,
          2.10000000e+01,   3.96900000e+02,   5.64000000e+00],
       [  1.09590000e-01,   0.00000000e+00,   1.19300000e+01, ...,
          2.10000000e+01,   3.93450000e+02,   6.48000000e+00],
       [  4.74100000e-02,   0.00000000e+00,   1.19300000e+01, ...,
          2.10000000e+01,   3.96900000e+02,   7.88000000e+00]]), 'feature_names': array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], 
      dtype='|S7'), 'DESCR': "Boston House Prices dataset\n\nNotes

In [4]:
X=bean.data
print X

[[  6.32000000e-03   1.80000000e+01   2.31000000e+00 ...,   1.53000000e+01
    3.96900000e+02   4.98000000e+00]
 [  2.73100000e-02   0.00000000e+00   7.07000000e+00 ...,   1.78000000e+01
    3.96900000e+02   9.14000000e+00]
 [  2.72900000e-02   0.00000000e+00   7.07000000e+00 ...,   1.78000000e+01
    3.92830000e+02   4.03000000e+00]
 ..., 
 [  6.07600000e-02   0.00000000e+00   1.19300000e+01 ...,   2.10000000e+01
    3.96900000e+02   5.64000000e+00]
 [  1.09590000e-01   0.00000000e+00   1.19300000e+01 ...,   2.10000000e+01
    3.93450000e+02   6.48000000e+00]
 [  4.74100000e-02   0.00000000e+00   1.19300000e+01 ...,   2.10000000e+01
    3.96900000e+02   7.88000000e+00]]


In [5]:
y=bean.target
print y

[ 24.   21.6  34.7  33.4  36.2  28.7  22.9  27.1  16.5  18.9  15.   18.9
  21.7  20.4  18.2  19.9  23.1  17.5  20.2  18.2  13.6  19.6  15.2  14.5
  15.6  13.9  16.6  14.8  18.4  21.   12.7  14.5  13.2  13.1  13.5  18.9
  20.   21.   24.7  30.8  34.9  26.6  25.3  24.7  21.2  19.3  20.   16.6
  14.4  19.4  19.7  20.5  25.   23.4  18.9  35.4  24.7  31.6  23.3  19.6
  18.7  16.   22.2  25.   33.   23.5  19.4  22.   17.4  20.9  24.2  21.7
  22.8  23.4  24.1  21.4  20.   20.8  21.2  20.3  28.   23.9  24.8  22.9
  23.9  26.6  22.5  22.2  23.6  28.7  22.6  22.   22.9  25.   20.6  28.4
  21.4  38.7  43.8  33.2  27.5  26.5  18.6  19.3  20.1  19.5  19.5  20.4
  19.8  19.4  21.7  22.8  18.8  18.7  18.5  18.3  21.2  19.2  20.4  19.3
  22.   20.3  20.5  17.3  18.8  21.4  15.7  16.2  18.   14.3  19.2  19.6
  23.   18.4  15.6  18.1  17.4  17.1  13.3  17.8  14.   14.4  13.4  15.6
  11.8  13.8  15.6  14.6  17.8  15.4  21.5  19.6  15.3  19.4  17.   15.6
  13.1  41.3  24.3  23.3  27.   50.   50.   50.   2

In [6]:
bean.data[0]

array([  6.32000000e-03,   1.80000000e+01,   2.31000000e+00,
         0.00000000e+00,   5.38000000e-01,   6.57500000e+00,
         6.52000000e+01,   4.09000000e+00,   1.00000000e+00,
         2.96000000e+02,   1.53000000e+01,   3.96900000e+02,
         4.98000000e+00])

In [7]:
def load_boston():
    scaler = StandardScaler()
    boston = datasets.load_boston()
    X=boston.data
    y=boston.target
    X = scaler.fit_transform(X)
    return train_test_split(X,y)

In [8]:
print train_test_split(X,y)

[array([[  2.00900000e-02,   9.50000000e+01,   2.68000000e+00, ...,
          1.47000000e+01,   3.90550000e+02,   2.88000000e+00],
       [  2.73397000e+00,   0.00000000e+00,   1.95800000e+01, ...,
          1.47000000e+01,   3.51850000e+02,   2.14500000e+01],
       [  1.35540000e-01,   1.25000000e+01,   6.07000000e+00, ...,
          1.89000000e+01,   3.96900000e+02,   1.30900000e+01],
       ..., 
       [  1.77800000e-02,   9.50000000e+01,   1.47000000e+00, ...,
          1.70000000e+01,   3.84300000e+02,   4.45000000e+00],
       [  2.61690000e-01,   0.00000000e+00,   9.90000000e+00, ...,
          1.84000000e+01,   3.96300000e+02,   1.17200000e+01],
       [  2.18700000e-02,   6.00000000e+01,   2.93000000e+00, ...,
          1.56000000e+01,   3.93370000e+02,   5.03000000e+00]]), array([[  1.27440000e-01,   0.00000000e+00,   6.91000000e+00, ...,
          1.79000000e+01,   3.85410000e+02,   4.84000000e+00],
       [  1.71420000e-01,   0.00000000e+00,   6.91000000e+00, ...,
       

In [9]:
X_train, X_test, y_train, y_test = load_boston()

<b>Fitting a Linear Regression

In [10]:
clf = LinearRegression()
clf.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [11]:
print('Coefficients: \n', clf.coef_)

('Coefficients: \n', array([-0.86812895,  1.02620075, -0.0573222 ,  0.35087623, -2.22237407,
        2.65546141,  0.13389253, -3.23133282,  2.61354119, -2.03017702,
       -1.8636904 ,  0.90545084, -3.69505876]))


In [12]:
print('INtercept: \n', clf.intercept_)

('INtercept: \n', 22.279770417467542)


<b>Making Prediction

In [13]:
zip (y_test, clf.predict(X_test))

[(27.5, 19.870180374466383),
 (27.899999999999999, 31.61648909661632),
 (32.5, 31.231178813421579),
 (7.5, 13.12852501154179),
 (19.600000000000001, 22.981587157326686),
 (12.1, 18.234123086748149),
 (23.199999999999999, 17.128981202252294),
 (22.5, 22.244644021347753),
 (30.5, 29.775087679324749),
 (22.5, 17.864359749340558),
 (21.100000000000001, 22.231176267575627),
 (22.199999999999999, 19.304542294137253),
 (20.5, 20.237827784616918),
 (14.300000000000001, 14.074622093410799),
 (20.0, 15.903814031318614),
 (28.699999999999999, 30.886818303211953),
 (24.800000000000001, 24.695406990929616),
 (22.899999999999999, 20.381757655292631),
 (50.0, 42.759837220476555),
 (13.9, 13.616458726273009),
 (7.4000000000000004, 5.766005574371416),
 (22.600000000000001, 18.867258844083338),
 (28.600000000000001, 27.845363887778092),
 (22.699999999999999, 20.55655792436632),
 (12.699999999999999, 17.91019434277332),
 (27.100000000000001, 19.238556876385147),
 (32.399999999999999, 34.533963065040886),

In [14]:
mean_squared_error(y_test, clf.predict(X_test))

26.014099624354014

In [15]:
r2_score(y_test, clf.predict(X_test))

0.72240386038666415

<b>From this it is seen that
<li>MSE is 28.1638093614798
<li>r2_score is 0.7070490824283745

In [16]:
#Now we will see Ridge linear regression 
#lets assign a value to alpha and instiantiate new object as aplpha=0.01
r = Ridge(alpha=.01)
r.fit(X_train, y_train)

Ridge(alpha=0.01, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [22]:
#lets compare the real value to the predicted values and create a tuple of the observation by Ridge model
zip(y_test,r.predict(X_test))

[(27.5, 19.869806569982998),
 (27.899999999999999, 31.615546590919205),
 (32.5, 31.230498385254847),
 (7.5, 13.128697859934043),
 (19.600000000000001, 22.980885064466207),
 (12.1, 18.234022401242871),
 (23.199999999999999, 17.128932258068403),
 (22.5, 22.24465409887366),
 (30.5, 29.775531421217572),
 (22.5, 17.864221052547993),
 (21.100000000000001, 22.230766261930377),
 (22.199999999999999, 19.304817077069856),
 (20.5, 20.236456526300866),
 (14.300000000000001, 14.074871670358846),
 (20.0, 15.903769279166411),
 (28.699999999999999, 30.886989624300192),
 (24.800000000000001, 24.695645783641183),
 (22.899999999999999, 20.382999637096461),
 (50.0, 42.759673061823079),
 (13.9, 13.616827863080134),
 (7.4000000000000004, 5.7661633784771062),
 (22.600000000000001, 18.867388131884223),
 (28.600000000000001, 27.845118067300199),
 (22.699999999999999, 20.556941173935428),
 (12.699999999999999, 17.90998812075685),
 (27.100000000000001, 19.238817457895298),
 (32.399999999999999, 34.53345622771689

In [20]:
mean_squared_error(y_test, r.predict(X_test))

26.014092143724266

In [21]:
r2_score(y_test, r.predict(X_test))

0.72240394021237286

<b>r2_score and mse on the Ridge linear reggessor model is
<li>r2_score=0.72240394021237286
<li>MSE=26.014092143724266

In [23]:
#lets assign a value to alpha and instiantiate new object as aplpha=0.02
r = Ridge(alpha=.02)
r.fit(X_train, y_train)

Ridge(alpha=0.02, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [24]:
mean_squared_error(y_test, r.predict(X_test))

26.014085209407881

In [25]:
r2_score(y_test, r.predict(X_test))

0.72240401420837841

<b>when alpha is 0.02 r2_score and mse on the Ridge linear reggessor model is
<li>r2_score=0.72240401420837841
<li>MSE=26.014085209407881

<b>Conclusion
<li><b>After optimising the Ridge by changing its alpha value there is not significant change in MSE and r2_score